File: macos/Saathi/Sources/SaathiKit/VoiceLaneReport.swift
//
//  VoiceLaneReport.swift
//  SaathiKit
//
//  "If I talk to Saathi, where does my voice go?"
//
//  The sibling of ProviderReport, and asked for the same reason: a companion you speak to is
//  handling the most identifying thing you produce. The answer is not the same as the provider
//  answer, and that is the point of having it separately — in the `chain` lane your VOICE never
//  leaves the machine even when the provider is a cloud one, because speech-to-text and
//  text-to-speech both run on-device and only the transcript is sent. Someone deciding whether to
//  narrate what they are struggling with deserves to know that distinction exists.
//
//  Pure, like ProviderReport: the resolved configuration in, the same bytes out on either platform,
//  so `scripts/check-parity.sh` can diff the macOS and Windows clients against each other. Anything
//  that depends on this particular machine — which microphone, whether a local model is actually
//  running — belongs behind a flag, the way `provider --probe` already is.
//

import Foundation
import SaathiContract

public enum VoiceLaneReport {

    /// Deterministic. Same configuration in, same bytes out, on either platform.
    public static func describe(_ configuration: SaathiConfiguration) -> String {
        let row = configuration.providerRow
        let chosenExplicitly = configuration.provider != nil
        let model = configuration.resolvedModel.isEmpty ? "chosen by the backend" : configuration.resolvedModel

        var lines: [String] = []
        lines.append("provider: \(row.kind.rawValue)\(chosenExplicitly ? "" : "  (default — nothing configured)")")
        lines.append("  \(laneSummary(row.voice))")
        lines.append("")
        lines.append("  lane       \(row.voice.rawValue)")
        lines.append("  speech in  \(speechIn(row))")
        lines.append("  thinking   \(model) @ \(configuration.resolvedProviderBaseURL)")
        lines.append("  speech out \(speechOut(row))")
        lines.append("  your voice \(voicePrivacy(row))")
        lines.append("  turn       \(turnShape(row.voice))")
        return lines.joined(separator: "\n")
    }

    private static func laneSummary(_ lane: VoiceLane) -> String {
        switch lane {
        case .realtime:
            return "One open connection carries speech in and speech out — the companion can be interrupted mid-sentence."
        case .chain:
            return "Speech in, think, speech out as three separate steps. Slower, and it works with any model."
        }
    }

    /// In the chain lane both ends are on-device, whoever the provider is. That is not a fallback —
    /// it is the reason the chain lane is worth having at all.
    private static func speechIn(_ row: SaathiProvider) -> String {
        row.voice == .realtime ? "\(row.kind.rawValue), over the open connection" : "on this machine"
    }

    private static func speechOut(_ row: SaathiProvider) -> String {
        row.voice == .realtime ? "\(row.kind.rawValue), over the open connection" : "on this machine"
    }

    private static func voicePrivacy(_ row: SaathiProvider) -> String {
        if row.voice == .realtime { return "leaves this machine as audio" }
        if row.sendsDataOffMachine { return "stays on this machine — only the transcript is sent" }
        return "stays on this machine"
    }

    private static func turnShape(_ lane: VoiceLane) -> String {
        switch lane {
        case .realtime: return "one connection, no round trip per step"
        case .chain: return "three steps per turn, so expect a pause before Saathi answers"
        }
    }
}

File: windows/src/Saathi.Core/VoiceLaneReport.cs
//
//  VoiceLaneReport.cs
//  Saathi.Core
//
//  "If I talk to Saathi, where does my voice go?"
//
//  The C# half of a report that must come out byte-identical to the Swift one — see
//  macos/Saathi/Sources/SaathiKit/VoiceLaneReport.swift, and scripts/check-parity.sh, which diffs
//  the two clients running side by side. Two clients that disagree about where a learner's voice
//  goes would be worse than either of them being wrong on its own.
//
//  The interesting line is `your voice`. In the chain lane speech-to-text and text-to-speech both
//  run on-device, so even with a cloud provider doing the thinking the learner's AUDIO never leaves
//  the machine — only the transcript does. That is a materially different promise from the realtime
//  lane, and it is the reason this report exists separately from ProviderReport.
//
//  Windows has no voice implementation yet: this reports the lane the contract assigns, which is
//  real and checkable, and the WPF shell will implement it behind the same contract. Reporting a
//  lane the platform cannot yet run is deliberate — it is what makes the gap visible rather than
//  silently absent.
//

using Saathi.Contract;

namespace Saathi.Core;

public static class VoiceLaneReport
{
    /// <summary>Deterministic. Same configuration in, same bytes out, on either platform.</summary>
    public static string Describe(SaathiConfiguration configuration)
    {
        var row = configuration.ProviderRow;
        var chosenExplicitly = configuration.Provider is not null;
        var model = configuration.ResolvedModel.Length == 0 ? "chosen by the backend" : configuration.ResolvedModel;

        var lines = new List<string>
        {
            $"provider: {row.Kind.ToString().ToLowerInvariant()}{(chosenExplicitly ? "" : "  (default — nothing configured)")}",
            $"  {LaneSummary(row.Voice)}",
            "",
            $"  lane       {row.Voice.ToString().ToLowerInvariant()}",
            $"  speech in  {SpeechEnd(row)}",
            $"  thinking   {model} @ {configuration.ResolvedProviderBaseUrl}",
            $"  speech out {SpeechEnd(row)}",
            $"  your voice {VoicePrivacy(row)}",
            $"  turn       {TurnShape(row.Voice)}",
        };
        return string.Join("\n", lines);
    }

    private static string LaneSummary(VoiceLane lane) => lane switch
    {
        VoiceLane.Realtime =>
            "One open connection carries speech in and speech out — the companion can be interrupted mid-sentence.",
        _ => "Speech in, think, speech out as three separate steps. Slower, and it works with any model.",
    };

    /// <summary>In the chain lane both ends are on-device, whoever the provider is. That is not a
    /// fallback — it is the reason the chain lane is worth having at all.</summary>
    private static string SpeechEnd(SaathiProvider row) =>
        row.Voice == VoiceLane.Realtime
            ? $"{row.Kind.ToString().ToLowerInvariant()}, over the open connection"
            : "on this machine";

    private static string VoicePrivacy(SaathiProvider row)
    {
        if (row.Voice == VoiceLane.Realtime) return "leaves this machine as audio";
        if (row.SendsDataOffMachine) return "stays on this machine — only the transcript is sent";
        return "stays on this machine";
    }

    private static string TurnShape(VoiceLane lane) => lane switch
    {
        VoiceLane.Realtime => "one connection, no round trip per step",
        _ => "three steps per turn, so expect a pause before Saathi answers",
    };
}
